### ISSUE TITLE
Explicit exporting of symbols

### ISSUE THREAD

**Issue by frangio** [ethereum/solidity#6482]
Currently Solidity exports all symbols in a file, including those that were `import`ed from other files. Has there been discussion around changing this to an explicit exports model? It would be nice to be able to control what gets exported, to modularize an implementation while keeping that an implementation detail.

**Comment by chriseth** [comment]
Which syntax would you like to suggest?

**Comment by frangio** [comment]
I would go with a subset of JavaScript's syntax.

1) Exporting a symbol that is being defined.
```
export contract Foo { }
```

2) Exporting previously defined or imported symbols.
```
export { Bar, Qux };
```

Symbols can be contracts, interfaces, or libraries.

**Comment by chriseth** [comment]
What about backwards compatibility? Export everything if no `export` keyword is used in the whole file? What about exporting symbols imported from other files?

**Comment by frangio** [comment]
I think this should be a breaking change eventually: nothing should be exported unless explicitly requested. However, to ease transition, the semantics you describe are a good idea. Implicit exports could be deprecated some time after, and then removed in the next breaking version bump.

Exporting imported symbols is covered in part 2 of the syntax I suggested above.

**Comment by chriseth** [comment]
Sounds good! Of course this has to be a breaking change. I'm adding it to the 0.6.0 discussion.

**Comment by chriseth** [comment]
Especially with functions and other symbols at file-level, I think we should reconsider this.

I really like the `export { symbolA, symbolB }` syntax with the following semantics:

 - if there is no `export` statement, export all symbols can be imported into other files (and will be with a "simple" import statement) - this is the current behaviour
 - mulitiple `export` statements are disallowed
 - if `export {A, B, C}` is specified, only the specified symbols are accessible from outside, regardless of the way the import is specified. Whether the specified symbols are declared in the file itself or in turn imported is irrelevant.

This mechanism allows people to cleanly specified the external interface of a file at a single place.

**Comment by frangio** [comment]
What is the argument for disallowing multiple `export` statements?

**Comment by chriseth** [comment]
@frangio I don't see a big benefit and the downside is that you might not see the second export statement and think that the identifiers in the first is all that is exported.

**Comment by chriseth** [comment]
Discussion: Main question here is if we want to have a statement that lists exports or if we want to mark individual declarations to be exported. Another question is if imports should be re-exported by default or not. And then of course there is the question if this feature is needed at all.

**Comment by frangio** [comment]
Imports should not be re exported by default. One of the motivations for this feature is that tools will generate compilation artifacts for all of the "exported" contracts contained in the solc output. If importing contracts from a smart contract library this can result in a lot of compilation artifacts the user doesn't care about, because everything that was imported will now be exported.

**Comment by chriseth** [comment]
My main point for re-exporting is that in a file that has `import "A.sol" as A;`, the expression `A.x` should be valid whenever `x` is valid everywhere inside `A.sol`. I don't think accessibility should be convoluted with what artefacts are generated by the compiler. Every symbol still has a unique point of declaration and that is how the artefacts are determined.

**Comment by frangio** [comment]
> the expression `A.x` should be valid whenever `x` is valid everywhere inside `A.sol`.

Why is this? What this issue proposes is that not everything in scope inside `A.sol` should be exported. So `A.x` should only be valid if `x` is exported from `A`. 

>Every symbol still has a unique point of declaration and that is how the artefacts are determined.

Last time I checked this was not true of Truffle or Buidler, because artifacts are determined by contract name, although Buidler is planning to change this.

**Comment by chriseth** [comment]
> > the expression `A.x` should be valid whenever `x` is valid everywhere inside `A.sol`.
> 
> Why is this? What this issue proposes is that not everything in scope inside `A.sol` should be exported. So `A.x` should only be valid if `x` is exported from `A`.

Sure, I was talking about the current situation or the possible identifiers you can export explicitly.

> > Every symbol still has a unique point of declaration and that is how the artefacts are determined.
> 
> Last time I checked this was not true of Truffle or Buidler, because artifacts are determined by contract name, although Buidler is planning to change this.

I don't understand how this would contradict my statement.

**Comment by frangio** [comment]
In some cases you can end up with different contracts in different scopes having the same name. For example, if one of your dependencies vendored `Ownable` and you also have `Ownable` through OpenZeppelin. This is fine with Solidity because it can tell them apart, but since they are all "exported" in solc's JSON output, and these tools create artifacts in the file system for all exported symbols, they will only create a single `Ownable.json` file (probably arbitrarily chosen betwen the two).

I hope this was a better example of the benefits I see from making exports explicit.

You could see this as a problem with the tools, and indeed Buidler is going to fix it, but I'd say the root cause is the namespace pollution caused by the common global scope.

**Comment by frangio** [comment]
Hm, I realize now that making exports explicit would not solve the issue I described above with the JSON artifact files... I mixed things up.

The example with a vendored `Ownable` could still be useful to analyze the feature. Since exports are implicit, if the user tries to combine those two dependencies in the same file, and `Ownable` is a transitive dependency, they would get a name clash between those two `Ownable` contracts that are actually defined in different places.
